lib/httputils/statconn.go

```diff
@@ -0,0 +1,93 @@
+package httputils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/netutil"
+	"github.com/VictoriaMetrics/metrics"
+)
+
+var (
+	dialsTotal *metrics.Counter
+	dialErrors *metrics.Counter
+	conns      *metrics.Counter
+
+	connReadsTotal   *metrics.Counter
+	connWritesTotal  *metrics.Counter
+	connReadErrors   *metrics.Counter
+	connWriteErrors  *metrics.Counter
+	connBytesRead    *metrics.Counter
+	connBytesWritten *metrics.Counter
+)
+
+var metricSetRegister sync.Once
+
+// GetStatDialFunc returns dial function that supports DNS SRV records,
+// and register stats metrics for conns.
+func GetStatDialFunc(metricPrefix string) func(ctx context.Context, network, addr string) (net.Conn, error) {
```

Comment:
I don't think this approach will work if application needs to have 2 or more statDialers. For example, if vmalert will try re-using this for datasource and remotewrite urls.

Reply:
fixed, please check https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6299/commits/10b3bba638c6a8bdf07976cd7f394caa7b333539.